### PR TITLE
Warning and Error consolidation CLI options

### LINKED ISSUE #364: Show all warnings
### OpenTF Version

```shell
1.0
```

### Use Cases

When upgrading a provider eg the aws provider, many warnings can occur, but only the first of them is shown. Eg 
```
Warning: Argument is deprecated
  on .terraform/modules/spa.s3_logs/main.tf line 1, in resource "aws_s3_bucket" "this":
   1: resource "aws_s3_bucket" "this" {
Use the aws_s3_bucket_acl resource instead

(and 21 more similar warnings elsewhere)
```
There is currently no option to show those similar warnings. All these warnings could be in different files, so we are forced to first address the only one shown, then re-run the terraform plan. Sometimes this requires many cycles. TEDIOUS!

### Attempted Solutions

Run `terraform validate`. 

However this does not seem to catch these warnings, possibly because the warnings are not about invalid code, just deprecation warnings. 

### Proposal

Several possibilities: 
- Add a switch on the plan and apply commands, eg `--show-all-warnings`. This switch should work even if the output is json. 
- Add a new command, `show-warnings`: if the users runs an apply or a plan and sees deprecation warnings, then they could run this new command to get the details. This makes sense if there are steps of the plan / apply can be skipped when the user just wants these (eg one would expect that it would be sufficient for terraform to parse the code, and ask the provider for any warnings about it, without actually having to check the cloud resources status, which could save a ton of time)

### References

- The issue is mentioned in a few places but the biggest discussion I found is https://discuss.hashicorp.com/t/show-all-warnings/5606/18. This demonstrates that HC is not interested in implementing this much-needed and almost trivial capability.

### LINKED ISSUE #1480: Merge duplicate diagnostics to avoid diagnostics flood
### OpenTofu Version

```shell
any
```

### Use Cases

When I use the template to run a plan 
```
output "out" {
  value = [for i in range(50): length([for i in range(50):bad()])]
}
```
I get 2500 same diagnostics :
image

### Attempted Solutions

merge them
or wait upstream(hcl) fix it

### Proposal

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by cam72cam**
Originally started by @Hefeweizen in #399, rebased and continued by @cam72cam.  Additional tests added by @Ericwww.

Resolves #1480
Resolves #364 

## Target Release

1.9.0

## Checklist

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.

### Go checklist

 

- [x] I have run golangci-lint on my change and receive no errors relevant to my code.
- [x] I have run existing tests to ensure my code doesn't break anything.
- [x] I have added tests for all relevant use cases of my code, and those tests are passing.
- [x] I have only exported functions, variables and structs that should be used from other packages.
- [x] I have added meaningful comments to all exported functions, variables, and structs.

### Website/documentation checklist

- [ ] I have locally started the website as [described here](https://github.com/opentofu/opentofu/blob/main/website/README.md) and checked my changes.

### CODE REVIEW

File: internal/tfdiags/consolidate_warnings.go [hunk 1]
```diff
@@ -7,112 +7,123 @@ package tfdiags
...
+		panic(fmt.Sprintf("Invalid diagnostic severity: %#v", wg.Severity()))
+	}
+
 	switch extraCount {
 	case 1:
-		msg = "(and one more similar warning elsewhere)"
+		msg = fmt.Sprintf("(and one more similar %s elsewhere)", diagType)
 	default:
-		msg = fmt.Sprintf("(and %d more similar warnings elsewhere)", extraCount)
+		msg = fmt.Sprintf("(and %d more similar %ss elsewhere)", extraCount, diagType)
```
Comment (Ericwww) [inline internal/tfdiags/consolidate_warnings.go:126]: some typo here? 

`%ss` => `%s`

Comment (cam72cam) [reply internal/tfdiags/consolidate_warnings.go:126]: I don't think so:

error -> errors
warning -> warnings

File: internal/tfdiags/consolidate_warnings.go
Comment (Ericwww) [inline internal/tfdiags/consolidate_warnings.go:1]: Maybe the filename should change also.

The function in this file is not only consolidating warnings now  :)

File: internal/tfdiags/consolidate_warnings.go (hunk 1 above)
Comment (Ericwww) [reply internal/tfdiags/consolidate_warnings.go:126]: Ah, I got you.

File: internal/tfdiags/consolidate_warnings.go
Comment (cam72cam) [reply internal/tfdiags/consolidate_warnings.go:1]: Git does not pick up on the rename properly when I do so.  I can do that in a follow-up PR to separate the changes.

File: internal/tfdiags/consolidate_warnings_test.go [hunk 2]
```diff
@@ -96,7 +96,7 @@ func TestConsolidateWarnings(t *testing.T) {
 
 	// We're using ForRPC here to force the diagnostics to be of a consistent
 	// type that we can easily assert against below.
-	got := diags.ConsolidateWarnings(2).ForRPC()
+	got := diags.Consolidate(2, Warning).ForRPC()
```
Comment (ollevche) [inline internal/tfdiags/consolidate_warnings_test.go:99]: nit: I would prefer constants for `1` and `2` when calling `Consolidate`. It isn't clear to me why exactly this number should be passed.

**Review by ollevche — APPROVED** [review]
LGTM! Left a minor nit comment (feel free to ignore)

File: internal/tfdiags/consolidate_warnings_test.go (hunk 2 above)
Comment (cam72cam) [reply internal/tfdiags/consolidate_warnings_test.go:99]: Other than this test, it's always 1.  The function was originally built to support consolidation at different count thresholds.  I see this as a legacy option that I didn't want to touch in this PR, even though the arg should probably be removed.

**Pull request merged**